### PR TITLE
[WIP] #557 snap to vertex and edge

### PR DESCRIPTION
**Pull request by ravraponi**
I have added new default options for snap-to.  Currently the snap-to option defaults to true with a default snapBuffer of 15 pixels. 
This enables the user to snap-to other point/line/polygon features within the draw tool.

All of the new options can be overridden.  

Example of overrriding the snapBuffer:

new MapboxDraw({
      displayControlsDefault: true,
      snapBuffer: 25
    });
    

The hover style for points and lines can be overridden as well. 
(Note: I did use the new circle stroke styling for the default snap-to styles so only latest Mapbox versions work with it. )

Example of overrriding the snap to over style:
    
new MapboxDraw({
  displayControlsDefault: true,
  snapTo: true,
  snapBuffer: 10,
  snapOverCircleStyle: {
  'id': 'gl-draw-circle-snap',
  'type': 'circle',
  'paint': {
    'circle-radius': 4,
    'circle-color': '#FF0',
    'circle-stroke-width' : 1,
    'circle-stroke-color' :'#000'
  },
  'filter': ['all', ["==", "id", ""]],
  'source': 'test-data'
  },
  snapOverLineStyle: {
  'id': 'gl-draw-line-snap',
  'type': 'line',
  'layout': {
    'line-cap': 'round',
    'line-join': 'round'
  },
  'paint': {
    'line-color': '#00F',
    'line-width': 2
  },
  'filter': ['all', ["==", "id", ""]],
  'source': 'test-data'
  }
});

Last but not least is controlling what features can be snapped-to, this is done by an array of style ids.
This enables the user to snap to features outside the drawing tool.

Example:

new MapboxDraw({
  displayControlsDefault: true,
  snapTo: true,
  snapBuffer: 10,
  snapStyles: ["test-polygon", "test-string", "test-point"]
});

FYI: I have used some turf methods to make snap possible.

To do -  A few old tests have been broken and I have not created new tests for snap-to or real examples. More testing is always needed. Plus code cleanup to conform to this plugins code style.

My time availability has dried up.  

Does anyone have time to contributet to these outstanding items?

### CODE REVIEW

**Comment by mcwhittemore** [comment]
@ravraponi - thanks for this. I'm down a rabbit hole on another project right now, so I'm going to ping @davidtheclark to see if he has space to pick this up. If not, I'll take a look when I come above water.

Any other takers?

File: src/lib/snap_to.js [hunk 1]
```diff
@@ -0,0 +1,120 @@
+const StringSet = require('../lib/string_set');
+const coordEach = require('@turf/meta').coordEach;
+const point = require('turf-point');
+const pointOnLine  = require('@turf/point-on-line');
+const distance = require('@turf/distance');
```
Comment (andrewharvey) [inline src/lib/snap_to.js]: https://github.com/mapbox/cheap-ruler is ~31x faster than @turf/distance, and over small distances (which are what's a play here) the margin of error is negligible. Perhaps you could use cheap-ruler instead?

File: package.json [hunk 2]
```diff
@@ -64,16 +64,20 @@
...
+    "@turf/distance": "^3.7.3",
+    "@turf/point-on-line": "^3.7.3",
     "geojson-area": "^0.2.1",
     "geojson-extent": "^0.3.2",
     "geojson-normalize": "0.0.1",
     "geojsonhint": "^2.0.0beta",
     "hat": "0.0.3",
     "lodash.isequal": "^4.2.0",
     "point-geometry": "0.0.0",
+    "turf-meta": "^3.0.12",
```
Comment (andrewharvey) [inline package.json]: "Turf packages are now namespaced: please use @turf/meta instead"

File: package.json [hunk 3]
```diff
@@ -64,16 +64,20 @@
...
+    "turf-point": "^2.0.1",
```
Comment (andrewharvey) [inline package.json]: turf-point is now in @turf/helpers

**Comment by andrewharvey** [comment]
Do you think you could rebase this with master? There were a bunch of recent fixes, in particular https://github.com/mapbox/mapbox-gl-draw/commit/329cddd1844226ad5d9571a83692da6a01b5c676 which fixes the debug page.

**Comment by andrewharvey** [comment]
* From a usability perspective I think allowing the user the ability to override a snap by holding Ctrl is desirable.

* I think it would provide better user feedback if there was a circle shown at the snap to line location while hovering around the line.

* When drawing lines, snapping to the start of the line would be nice for those wanting to draw closed loops (as lines not polygons).

**Comment by andrewharvey** [comment]
>To do - A few old tests have been broken and I have not created new tests for snap-to or real examples. More testing is always needed. Plus code cleanup to conform to this plugins code style.
>
>My time availability has dried up.
>
>Does anyone have time to contributet to these outstanding items?

Since you indicated you don't have time to work on this, I'm going to give it a shot. I'm working at https://github.com/andrewharvey/mapbox-gl-draw/tree/snap-to-edge-and-vertex

File: src/lib/snap_to.js (hunk 1 above)
Comment (andrewharvey) [inline src/lib/snap_to.js]: working on this at https://github.com/andrewharvey/mapbox-gl-draw/tree/snap-to-edge-and-vertex

File: package.json (hunk 2 above)
Comment (andrewharvey) [inline package.json]: resolved at https://github.com/andrewharvey/mapbox-gl-draw/tree/snap-to-edge-and-vertex

File: package.json (hunk 3 above)
Comment (andrewharvey) [inline package.json]: resolved at https://github.com/andrewharvey/mapbox-gl-draw/tree/snap-to-edge-and-vertex

**Comment by ravraponi** [comment]
@andrewharvey 
Thanks for the assist! 
I am looking forward to to seeing the changes to making it faster and more precise.

**Comment by ravraponi** [comment]
@andrewharvey 
FYI:  I have merged your changes and added some precision changes of my own.

**Comment by andrewharvey** [comment]
@ravraponi Awesome.

I did notice that

- [ ] when drawing and it previews a snap before clicking to commit it, if you hit escape, the other feature is still styled like it's being snapped to

**Comment by davidtheclark** [comment]
@ravraponi for this and the other PR, please let us know with a comment when you've addressed all known issues and are ready for another review. We don't want to re-review prematurely, so this will help us know when it's the right time.

**Comment by mcwhittemore** [comment]
@ravraponi it looks like travis didn't get your latest push.

Is this ready for review?

**Comment by ravraponi** [comment]
@mcwhittemore 
Still a WIP.  I submitted this PR since snapping seemed a hot item and I had hoped others would contribute on tests (never used tape before) instead of duplicating the work I have already done.  I am still adding code whenever I have free time but that is pretty limited.

**Comment by mcwhittemore** [comment]
@ravraponi - thanks again for this work and thanks for adding `[wip]` to the title that helps me a bunch.

**Comment by grundid** [comment]
Hi. This is great work! I would like to contribute to this PR and add snap functions to this project.
I have a merged branch at
https://github.com/geohub-plattform/mapbox-gl-draw/tree/snap-to-functions

However I have specific requirements to be able to draw alongside the raw map data. So I have rewritten most parts of the snap_to.js file. For example I would like to draw a track (line) that is exactly like the way on the map.

I have a demo for this at:
https://geohub-plattform.github.io/geohub-client/

For now I would like to ask on how to continue on this. Do you want to have this snap functions tightly integrated into the draw editor or can we add some events that would allow to have snap functions as a plugin?
The above work is just the beginning. I'm working on an editor that should allow users to create polygons and tracks based on existing map data. This means that the editor should also add any missing points between the last clicked point and the snap position. Of course this will only work within a small range due to performance issues.

**Comment by ravraponi** [comment]
@grundid 
Glad for the interest! It was nice to see your demo to see how you would like to use it.  Currently, the snapTo is integrated tightly into the drawing and I do not foresee that changing.  Most of what you want to do is possible using snap-to as is.
It looks like you found the MultiLineString bug which I will look into very soon.
Also I plan to add some a snap custom function in the future as well.

Note:  I have just moved the snapOverStyles into theme.  Still need to update documentation.

**Comment by grundid** [comment]
Hi. I wanted to give you an update on my progress with snapping. You can find my current demo at: 
https://geohub-plattform.github.io/milestone1/

Here the editor already has some data preloaded (in gray) and you can use it to snap to this data. This also involves some routing logic. For this demo I've stripped all mapbox-gl-draw functions, so you can only draw lines.

**Comment by ravraponi** [comment]
Thanks for sharing what you have done. 
 I like the way it is working, and the look and feel of using it. 
What are your plans going forward?

**Comment by grundid** [comment]
I'm working on a geo online editor with this kind of snap functionality at its core. Also "undo" functions are quite important for my editor.

I'd be glad if we could integrate this functions into this draw tool, but this requires some restructuring of the current functionality.
I found out that mapbox-gl tiles contain optimized geo data and not the real data that was uploaded via geojson files. For this reason my editor keeps a copy of all the geo information that was part of the original geojson file. 

Since I'm not very familiar with this draw tool's code, I'm not sure how to restructure the code to implement "undo" functions and where to keep the original geo data for route calculation.

**Comment by mcwhittemore** [comment]
@grundid - Using a GeoJSON source is a good way to keep your data non-optimized, but you do not need to use MapboxDraw to do this. [Here is a link to the mapbox-gl docs for how to add your own GeoJSON layer](https://www.mapbox.com/mapbox-gl-js/api/#geojsonsource).

**Comment by mcwhittemore** [comment]
With the advent of custom modes (#671), Mapbox Draw functionality is locked for v1. I'd love to see these changes written as a user land mode. If that happens, please feel free to open a PR adding the mode to [our custom mode list](https://github.com/mapbox/mapbox-gl-draw/blob/master/src/modes/README.md#available-custom-modes).

**Pull request closed without merging**